modules/storage-api/src/main/java/org/apache/ignite/internal/storage/PartitionStorage.java

```diff
@@ -128,6 +143,66 @@
      */
     Cursor<DataRow> scan(Predicate<SearchRow> filter) throws StorageException;
 
+    /**
+     * Scans the partition and returns a cursor of values in at the given timestamp.
+     *
+     * @param keyFilter Key filter. Binary rows passed to the filter may or may not have a value, filter should only check keys.
+     * @param timestamp Timestamp
+     * @return Cursor.
+     */
+    default Cursor<BinaryRow> scan(Predicate<BinaryRow> keyFilter, @Nullable Timestamp timestamp) {
+        throw new UnsupportedOperationException("scan");
+    }
+
+    /**
+     * Exception class that describes the situation where two independant transactions attempting to write values for the same key.
+     */
+    class TxIdMismatchException extends RuntimeException {
+    }
```

Comment:
Let's move to a separate high-level class (or make is static at least).

Reply:
Why does not inherit `IgniteException`?